internal/engine/wazevo/engine.go

```diff
@@ -227,27 +226,81 @@ func (e *engine) compileModule(ctx context.Context, module *wasm.Module, listene
 	needCallTrampoline := callTrampolineIslandSize > 0
 	var callTrampolineIslandOffsets []int // Holds the offsets of trampoline islands.
 
-	for i := range module.CodeSection {
-		if wazevoapi.DeterministicCompilationVerifierEnabled {
-			i = wazevoapi.DeterministicCompilationVerifierGetRandomizedLocalFunctionIndex(ctx, i)
-		}
+	type CompiledLocalFuncResult struct {
+		Body             []byte
+		RelsPerFunc      []backend.RelocationInfo
+		IDX              wasm.Index
+		SourceOffsetInfo []backend.SourceOffsetInfo
+	}
+
+	compiledFuncs := make([]CompiledLocalFuncResult, len(module.CodeSection))
+
+	workers := runtime.GOMAXPROCS(0)
+
+	wg := sync.WaitGroup{}
+	wg.Add(workers)
+
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(nil)
+
+	sections := sequence(len(module.CodeSection))
+
+	for range workers {
+		go func() {
+			defer wg.Done()
+
+			ssaBuilder := ssa.NewBuilder()
+			machine := newMachine()
+			fe := frontend.NewFrontendCompiler(module, ssaBuilder, &cm.offsets, ensureTermination, withListener, needSourceInfo)
+
+			for i := range sections {
+				if err := ctx.Err(); err != nil {
+					// Compilation canceled!
+					return
+				}
+
+				if wazevoapi.DeterministicCompilationVerifierEnabled {
+					i = wazevoapi.DeterministicCompilationVerifierGetRandomizedLocalFunctionIndex(ctx, i)
+				}
```

Comment:
worth noticing that enabling this flag in `debug_options.go` give us further information about what could be happening. Do notice that this fails even without pushing `be` outside the loop.

```
❯ go run ./cmd/wazero run ../repro/redis-installer.wasm
BUG: Deterministic compilation failed for function[3/55834].internal_abi.__IntArgRegBitmap_.Set at scope="SSA".

This is mostly due to (but might not be limited to):
        * Resetting ssa.Builder, backend.Compiler or frontend.Compiler, etc doens't work as expected, and the compilation has been affected by the previous iterations.
        * Using a map with non-deterministic iteration order.
        
[... more details ...]
```

Reply:
I don't exclude that the check currently might not be concurrency-safe :)

Reply:
Yeah, when I try and enable that check, I get a concurrency error:

```
fatal error: concurrent map read and map write
```

Setting up the verifierState. Unsure how you got it setup to verify anything :\

Reply:
Okay sometimes it gives me that error and sometimes it gives me your error.

If I run with the verifier against main, it just hangs and never completes.

Hmmm.

Reply:
@davidmdm if the problem is deterministic builds being prevented due to the random order of the sequence generate I might have a few solutions you can try via seeding.